Allow tests in examples to use cuda or fp16,if they are available

Joel-hanson opened the pull request:
The tests in examples didn't use the Cuda or fp16 even if they were available.

- The text classification example (`run_glue.py`) didn't use the fp16 even if it was available but
  the device was take based on the availability(Cuda/CPU).
- The language-modeling example (`run_language_modeling.py`) was having `--no_cuda` argument which
  made the test work without Cuda. This example is having an issue when running with fp16
  thus it not enabled (got an assertion error for perplexity due to its higher value).
- The Cuda and fp16 is not enabled for question-answering example (`run_squad.py`) as it is having a
  the difference in the f1 score.
- The text-generation example (`run_generation.py`) will take the Cuda or fp16 whenever it is available.

Resolves some of #5057

Joel-hanson commented:
@sshleifer Hope you are doing well. sorry for the delay, I have created the PR for some related to the first issue.
> 1. They never use Cuda or fp16, even if they are available.

I have some doubts which had encountered when making this PR

---
**1.** As you had said there where some test which failed when enabling the Cuda or fp16
- The Cuda and fp16 are not enabled for question-answering example (`run_squad.py`) as it is having a difference in the f1 score.
- The language-modeling example (`run_language_modeling.py`) is having an issue when running with fp16.
---
**2.** I was not able to find the `test_hans.py` but was able to find a readme to run it. Is this intentional if not shall I have a `test_hans.py` file to run the same.

![Screenshot 2020-07-04 at 11 56 30 AM](https://user-images.githubusercontent.com/17215044/86506506-7c634200-bded-11ea-8b1a-a49fa617f8ea.png)
---
**3.** This is the list of tests which I got
```bash
$ ls -l examples/**/test*.py

examples/bert-loses-patience/test_run_glue_with_pabee.py
examples/seq2seq/bertabs/test_utils_summarization.py
examples/seq2seq/test_seq2seq_examples.py
examples/test_examples.py
examples/token-classification/test_ner_examples.py
```
I was not able to find the `test_summarization_examples.py` and `test_t5_examples.py`. I think I am doing something wrong.

---
**4.** I can have made the PR for some tests only, can do the same for others if the current PR satisfies your requirement.

sshleifer commented:
1) noted
2)test_hans.py would be nice, but can be added in a separate PR.
3) You are all set, nothing is wrong. 
4) Would love to see the current PR!

Thanks!

Joel-hanson commented:
>    2)test_hans.py would be nice, but can be added in a separate PR.

The test_hans.py was created once via the PR https://github.com/huggingface/transformers/pull/2239 but I think somehow or for some reason, it was removed

last interaction with the file was in the PR https://github.com/huggingface/transformers/pull/4213

sshleifer commented on examples/text-classification/run_glue.py:
```diff
@@ -85,6 +86,9 @@ def main():
             f"Output directory ({training_args.output_dir}) already exists and is not empty. Use --overwrite_output_dir to overcome."
         )
 
+    if "cuda" not in str(training_args.device) or not is_apex_available():
```
Can we just change the test code? Or is this fixing a real bug?

Joel-hanson replied:
@sshleifer I wasn't sure whether this is a good solution but I have added this for certain reasons.
1.  For the apex to work we need to have cuda. So when was implementing this I had a scenario where the apex was installed but the GPU is not enabled thus I added both the checks. This checks for the cuda first before checking if apex is available or not.
---
2. I wasn't sure if the `training_args.device` will give `cuda` or `cuda:0` when there are multiple GPUs thus I have added a check to know if the `training_args.device` has a cuda. 
---
> Did you also add multi GPU support? If so you need to test it on a machine/show that it works, and it should probably be a separate PR.
3. I didn't find a way to test this on multiple GPU. I will try my best to find a way to do this. But in `examples/text-generation/run_generation.py` I have added `local_rank` because I need to make every test file to have consistent arguments. This doesn't mean it adds a multi GPU support, I can remove this as it is confusing.

I am pretty sure this might be a wrong implementation please correct me if any of the above-mentioned points are not correct and if there is any other best alternative to do this.

sshleifer replied:
Lets check if cuda is available in `test_examples.py`, and if it is pass the fp16 argument.
Otherwise, don't pass the fp16 argument.

sshleifer commented:
also I missed this earlier, but what was the issue with run_language_modeling.py? Can you get a traceback?

Joel-hanson commented:
> also I missed this earlier, but what was the issue with run_language_modeling.py? Can you get a traceback?

This is the short long and I have attached the pull log with this comment.
```log
        with patch.object(sys, "argv", testargs):
            result = run_language_modeling.main()
>           self.assertLess(result["perplexity"], 35)
E           AssertionError: 36.684365356893885 not less than 35
```
[test.log](https://github.com/huggingface/transformers/files/4955269/test.log)

Joel-hanson commented:
@sshleifer Hope you are doing well.

Is this PR ok or Should I split this into multiple smaller PRs to make things easier?

Joel-hanson commented:
Thanks for the review @sshleifer, I have updated the code accordingly and regarding the new command line args, I have removed the args which were related to the multi-GPU support as we can make that a separate PR. 

**Any Idea why the test `run_tests_tf` and `run_tests_torch_and_tf` failing? Is it related to my change?**

sshleifer commented on examples/text-generation/run_generation.py:
```diff
@@ -186,11 +186,20 @@ def main():
     args.device = torch.device("cuda" if torch.cuda.is_available() and not args.no_cuda else "cpu")
     args.n_gpu = 0 if args.no_cuda else torch.cuda.device_count()
 
+    logger.warning(
```
on line 213 you need to add code to actually use `args.fp16`
```python
if args.fp16: model.half()
```

Joel-hanson replied:
@sshleifer I have made this change in `run_generation.py`, But I have a doubt `run_glue.py` should also have this change right? 
But when I tried I got an error with a message as follows:
`msg = 'Found param distilbert.embeddings.word_embeddings.weight with type torch.cuda.HalfTensor, expected torch.cuda.FloatTe...alize, you do not need to call .half() on your model\nbefore passing it, no matter what optimization level you choose.'`

sshleifer replied:
you did the right thing!

LysandreJik commented on examples/test_examples.py:
```diff
@@ -51,6 +53,10 @@ def get_setup_file():
     return args.f
 
 
+def is_cuda_and_apex_avaliable():
+    return torch.cuda.is_available() and is_apex_available()
+
+
```
This should also depend from the `USE_CUDA` flag, similarly to the rest of the test suite. Even if we decide to set `USE_CUDA` to `True` by default, setting `USE_CUDA` to `False` should result in the examples not using CUDA imo.

LysandreJik replied:
I think you can do so by adding the following line:

```py
from transformers.testing_utils import torch_device
```
and updating the statement to
```py
return torch.cuda.is_available() and is_apex_available() and torch_device == "cuda"
```

Joel-hanson replied:
Good catch @LysandreJik

LysandreJik commented on examples/test_examples.py:
```diff
@@ -131,7 +141,6 @@ def test_run_language_modeling(self):
             --do_train
             --do_eval
             --num_train_epochs=1
-            --no_cuda
```
Can update this to respect the `USE_CUDA` flag as well:

Something like 
```py
        testargs = """
            run_language_modeling.py
            --model_name_or_path distilroberta-base
            [...]
            --num_train_epochs=1
            {'--no_cuda' if torch_device != 'cuda' else ''}
            """.split()
```

Joel-hanson commented:
Thanks for the review @LysandreJik, I have updated the PR accordingly.

LysandreJik commented on examples/test_examples.py:
```diff
@@ -52,6 +53,15 @@ def get_setup_file():
+
+
+def clean_test_dir(path):
+    shutil.rmtree(path, ignore_errors=True)
```
You should import shutil for this to work (that's why the code quality fails)
